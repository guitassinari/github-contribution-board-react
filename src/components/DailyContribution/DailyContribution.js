import React from 'react'
import PropTypes from 'prop-types'
import Square from '../Square/Square.js'

const CONTRIBUTION_COLORS_SETTINGS = {
  NONE: { min: 0, max: 0, color: "#EBEDF0" },
  LOW: { min: 1, max: 2, color: "#9BE9A8"},
  MEDIUM: { min: 3, max: 5, color: "#40C463"},
  HIGH: { min: 6, max: 8, color: "#30A14E"},
  VERY_HIGH: { min: 9, max: Infinity, color: "#216E39"}
}

function getContributionBackgroundColor(contributionCount) {
  if(!contributionCount) {
    return CONTRIBUTION_COLORS_SETTINGS.NONE.COLOR
  }

  const allSettings = Object.values(CONTRIBUTION_COLORS_SETTINGS)
  const contributionIntensity = allSettings.find(settings => {
    return contributionCount >= settings.min &&
           contributionCount <= settings.max
  })

  return contributionIntensity.color
}

function DailyContribution({ count }) {
  const backgroundColor = getContributionBackgroundColor(count)

  return (
    <Square backgroundColor={backgroundColor} />
  )
}

DailyContribution.propTypes = {
  count: PropTypes.number.isRequired
}

DailyContribution.defaultProps = {
  count: 0
}

export default DailyContribution