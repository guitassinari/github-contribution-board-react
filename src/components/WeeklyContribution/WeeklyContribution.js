import React from 'react'
import PropTypes from 'prop-types'
import DailyContribution from '../DailyContribution/DailyContribution.js'
import styled from 'styled-components'

const ContributionsContainer = styled.div`
  display: grid;
  grid-gap: 3px;
  align-content: start;
`

function WeeklyContribution({ contributions }) {
  const filteredContributions = (contributions || []).slice(0, 7) 
  return (
    <ContributionsContainer>
      {filteredContributions.map((contributionCount, index) => <DailyContribution key={index} count={contributionCount} />)}
    </ContributionsContainer>
  )
}

WeeklyContribution.propTypes = {
  contributions: PropTypes.arrayOf(PropTypes.number)
}

WeeklyContribution.defaultProps = {
  contributions: [0, 0, 0, 0, 0, 0, 0]
}

export default WeeklyContribution