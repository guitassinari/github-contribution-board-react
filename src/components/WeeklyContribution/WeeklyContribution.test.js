import React from 'react';
import renderer from 'react-test-renderer'
import { shallow } from 'enzyme';
import WeeklyContribution from './WeeklyContribution';
import DailyContribution from '../DailyContribution/DailyContribution.js'

describe('WeeklyContribution', () => {
  test('Matches snapshot', () => {
    const component = renderer.create(<WeeklyContribution />);
    const tree = component.toJSON()
  
    expect(tree).toMatchSnapshot();
  });

  describe('when it receives null contributions', () => {
    const contributions = null 

    test('Matches snapshot', () => {
      const component = renderer.create(<WeeklyContribution contributions={contributions} />);
      const tree = component.toJSON()
    
      expect(tree).toMatchSnapshot();
    });

    test('renders no Daily Contributions', () => {
      const wrapper = shallow(<WeeklyContribution contributions={contributions} />);

      const renderedDailyContributions = wrapper.find(DailyContribution)
        
      expect(renderedDailyContributions.length).toEqual(0);
    })
  })

  describe('when it receives less than 7 contributions', () => {
    const contributions = [1, 2, 3, 4]

    test('Matches snapshot', () => {
      const component = renderer.create(<WeeklyContribution contributions={contributions} />);
      const tree = component.toJSON()
    
      expect(tree).toMatchSnapshot();
    });

    test('renders only the received contributions', () => {
      const wrapper = shallow(<WeeklyContribution contributions={contributions} />);

      const renderedDailyContributions = wrapper.find(DailyContribution)
        
      expect(renderedDailyContributions.length).toEqual(contributions.length);

      renderedDailyContributions.forEach((dailyContribution, index) => {
        expect(dailyContribution.props().count).toEqual(contributions[index])
      })
    })
  })

  describe('when it receives more than 7 contributions', () => {
    const contributions = [1, 2, 3, 4, 6, 7, 8, 9]

    test('Matches snapshot', () => {
      const component = renderer.create(<WeeklyContribution contributions={contributions} />);
      const tree = component.toJSON()
    
      expect(tree).toMatchSnapshot();
    });

    test('renders only the first seven contributions', () => {
      const wrapper = shallow(<WeeklyContribution contributions={contributions} />);

      const renderedDailyContributions = wrapper.find(DailyContribution)
        
      expect(renderedDailyContributions.length).toEqual(7);

      renderedDailyContributions.forEach((dailyContribution, index) => {
        expect(dailyContribution.props().count).toEqual(contributions[index])
      })
    })
  })
})
