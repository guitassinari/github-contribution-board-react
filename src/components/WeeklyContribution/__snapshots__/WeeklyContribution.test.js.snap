// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`WeeklyContribution Matches snapshot 1`] = `
<div
  className="sc-gtssRu rRoJ"
>
  <div
    className="sc-bdnylx gPrTLT"
  />
  <div
    className="sc-bdnylx gPrTLT"
  />
  <div
    className="sc-bdnylx gPrTLT"
  />
  <div
    className="sc-bdnylx gPrTLT"
  />
  <div
    className="sc-bdnylx gPrTLT"
  />
  <div
    className="sc-bdnylx gPrTLT"
  />
  <div
    className="sc-bdnylx gPrTLT"
  />
</div>
`;

exports[`WeeklyContribution when it receives less than 7 contributions Matches snapshot 1`] = `
<div
  className="sc-gtssRu rRoJ"
>
  <div
    className="sc-bdnylx bkFITR"
  />
  <div
    className="sc-bdnylx bkFITR"
  />
  <div
    className="sc-bdnylx fFbPkp"
  />
  <div
    className="sc-bdnylx fFbPkp"
  />
</div>
`;

exports[`WeeklyContribution when it receives more than 7 contributions Matches snapshot 1`] = `
<div
  className="sc-gtssRu rRoJ"
>
  <div
    className="sc-bdnylx bkFITR"
  />
  <div
    className="sc-bdnylx bkFITR"
  />
  <div
    className="sc-bdnylx fFbPkp"
  />
  <div
    className="sc-bdnylx fFbPkp"
  />
  <div
    className="sc-bdnylx eankiZ"
  />
  <div
    className="sc-bdnylx eankiZ"
  />
  <div
    className="sc-bdnylx eankiZ"
  />
</div>
`;

exports[`WeeklyContribution when it receives null contributions Matches snapshot 1`] = `
<div
  className="sc-gtssRu rRoJ"
/>
`;
